oauthproxy.go

```diff
@@ -589,6 +599,27 @@ func (p *OAuthProxy) GetRedirect(req *http.Request) (redirect string, err error)
 	return
 }
 
+// getRedirectFromForwardHeaders returns the redirect URL based on X-Forwarded-{Proto,Host,Port,Uri} headers
+func (p *OAuthProxy) getRedirectFromForwardHeaders(req *http.Request) string {
+	proto := fmt.Sprintf("%s://", util.GetRequestProto(req))
+
+	host, port := splitHostPort(util.GetRequestHost(req))
+	port = util.GetRequestPortFromHeader(req, port)
+	formedPort := fmt.Sprintf(":%s", port)
+
+	uri := util.GetRequestURI(req)
+
+	if port != "" && validOptionalPort(formedPort) {
+		port = formedPort
+	}
+
+	if strings.HasPrefix(uri, p.ProxyPrefix) {
+		uri = "/"
+	}
+
+	return fmt.Sprintf("%s%s%s%s", proto, host, port, uri)
```

Comment:
I think when you ditch the port logic, this becomes a lot simpler:

```
return fmt.Sprintf("%s://%s%s, util.GetRequestProto, util.GetRequestHost, uri
```

Probably still need the specialized `ProxyPrefix` overlap check.